### PR TITLE
[CPU] Add inplace and jit for ROPE

### PR DESCRIPTION
**Pull request by luo-cheng2021**
### Details:
 - *Add inplace and jit if possible*
 - *...*

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: src/plugins/intel_cpu/src/nodes/rope.cpp [hunk 1]
```diff
@@ -7,17 +7,281 @@
...
+    explicit jit_uni_rotary_kernel(const jit_rotary_compile_params& jcp) : ker_(nullptr), jcp_(jcp) {}
+    virtual ~jit_uni_rotary_kernel() {}
+    virtual void create_ker() = 0;
+    jit_rotary_compile_params jcp_;
+};
+
+#ifdef OPENVINO_ARCH_X86_64
+
+template <cpu_isa_t isa>
+struct jit_rotary_kernel : public jit_uni_rotary_kernel, public jit_generator {
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/rope.cpp:60]: Please move the kernel to dedicate folder: https://github.com/openvinotoolkit/openvino/tree/fc0b82c1be18f793b41e894981a061b90c31a1e1/src/plugins/intel_cpu/src/nodes/kernels/x64

File: src/plugins/intel_cpu/src/nodes/rope.cpp [hunk 2]
```diff
@@ -7,17 +7,281 @@
...
+        add(reg_cos, sizeof(float) * step);
+        add(reg_sin, sizeof(float) * step);
+    }
+    void rotary_interleave(size_t step) {
+        // for (size_t j = 0; i < rotary_dims; i += 2, j++) {
+        //     dst[i] = cos[j] * x[i] - sin[j] * x[i + 1];
+        //     dst[i + 1] = cos[j] * x[i + 1] + sin[j] * x[i];
+        // }
+        load(vmm_src0, reg_src, jcp_.src_prc, step, false);
+        lea(reg_tmp, ptr[reg_src + step * jcp_.src_prc.size()]);
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/rope.cpp:165]: Load emitter takes constant offset as second input param: https://github.com/openvinotoolkit/openvino/blob/fc0b82c1be18f793b41e894981a061b90c31a1e1/src/plugins/intel_cpu/src/emitters/plugin/x64/jit_load_store_emitters.cpp#L105. It means you can avoid multiple lea instruction by passing {reg_src, step * jcp_.src_prc.size()} to emit_code().

File: src/plugins/intel_cpu/src/nodes/rope.cpp [hunk 3]
```diff
@@ -7,17 +7,281 @@
...
 namespace node {
 
+struct jit_rotary_compile_params {
+    ov::element::Type src_prc;
+    ov::element::Type dst_prc;
+    size_t rotary_ndims;
+    bool interleave;
+    bool mix_cos_sin;
+};
+struct jit_rotary_call_args {
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/rope.cpp:39]: minor: please add const qualifier to the input pointer

File: src/plugins/intel_cpu/src/nodes/rope.cpp [hunk 4]
```diff
@@ -7,17 +7,281 @@
...
+            // higher 64bit/128 lane
+            //           16-18      20-22      24-26       28-30
+            static const uint64_t mask_zmm[] = {
+                0, 4, 1, 5, 2, 6, 3, 7
+            };
+            if (isa == cpu_isa_t::avx512_core) {
+                mov(reg_tmp, reinterpret_cast<uintptr_t>(mask_zmm));
+                uni_vmovups(vmm_idx, ptr[reg_tmp]);
+            }
+            auto half_rotary_ndims = jcp_.rotary_ndims / 2;
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/rope.cpp:96]: I am wondering what is the typical value for `rotary_ndims`?

File: src/plugins/intel_cpu/src/nodes/rope.cpp [hunk 5]
```diff
@@ -40,12 +307,34 @@ struct RoPE::RoPEExecutorRotateHalf : public RoPE::Executor {
...
+            jcp.interleave = false;
+            if (mayiuse(cpu_isa_t::avx512_core)) {
+                m_rotaryKernel.reset(new jit_rotary_kernel<cpu_isa_t::avx512_core>(jcp));
+            } else if (mayiuse(cpu_isa_t::avx2)) {
+                m_rotaryKernel.reset(new jit_rotary_kernel<cpu_isa_t::avx2>(jcp));
+            }
+            if (m_rotaryKernel)
+                m_rotaryKernel->create_ker();
+#endif
+        }
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/rope.cpp:328]: Logically it should be part of Executor ctor, then m_init field will not be needed

File: src/plugins/intel_cpu/src/nodes/rope.cpp [hunk 6]
```diff
@@ -221,6 +619,7 @@ void RoPE::initSupportedPrimitiveDescriptors() {
 
     auto rtPrecision = srcPrecision;
     auto CosSinPrecision = ov::element::f32;
+    bool can_inplace = true;
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/rope.cpp:622]: Making inplace optimization `true` by default is not very safe in context of future RoPE extension. Can we disable it by default and explicitly enable only for valid cases below?

File: src/plugins/intel_cpu/src/nodes/rope.cpp (hunk 1 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/rope.cpp]: Done.

File: src/plugins/intel_cpu/src/nodes/rope.cpp (hunk 2 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/rope.cpp]: Done.

File: src/plugins/intel_cpu/src/nodes/rope.cpp (hunk 3 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/rope.cpp]: Done.

File: src/plugins/intel_cpu/src/nodes/rope.cpp (hunk 5 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/rope.cpp]: Moved into ctor.

File: src/plugins/intel_cpu/src/nodes/rope.cpp (hunk 6 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/rope.cpp]: Done, only llama like will be enabled.

File: src/plugins/intel_cpu/src/nodes/rope.cpp (hunk 4 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/rope.cpp]: The vaule of some models:
llama2-7b/13b/mistral-7b: 128
chatglm2/3-6b: 64
qwen-7b: 128
gptj-6b: 64
gemm-7b: 192
gptneox-20b: 24

File: src/plugins/intel_cpu/src/nodes/rope.cpp [hunk 7]
```diff
@@ -151,27 +253,41 @@ struct RoPE::RoPEExecutorChatGLM : public RoPE::Executor {
...
+        jit_rotary_compile_params jcp;
+        jcp.src_prc = precision_of<T>::value;
+        jcp.dst_prc = precision_of<T>::value;
+        jcp.rotary_ndims = config.rotary_ndims;
+        jcp.interleave = false;
+        m_rotaryKernel = createJitKernel(jcp);
+    }
+
     void execute(dnnl::stream strm,
                  const op::internal::RoPE::Config& config,
```
[the remark this answers is not in the document]
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/rope.cpp]: Done.

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**